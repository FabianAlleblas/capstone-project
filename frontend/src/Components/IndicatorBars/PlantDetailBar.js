import PropTypes from 'prop-types'
import styled from 'styled-components'
import { SettingsIcon } from '../Icons'

PlantDetailBar.propTypes = {
  plant: PropTypes.object.isRequired,
  setShowCareSetting: PropTypes.func.isRequired,
}

export default function PlantDetailBar({ plant, setShowCareSetting }) {
  return (
    <Container>
      <BarWrapper>
        <Bar>
          <WaterIndicator
            data-testid="water-indicator"
            daysLeft={plant.daysLeft}
            waterInterval={plant.waterInterval}
          />
        </Bar>
        <CareIntervalWrapper>
          <Text>
            Water {plant.daysLeft}/{plant.waterInterval} days left
          </Text>
          <IconButton
            data-testid="water-button"
            name="water"
            onClick={openCareSettings}
          >
            <SettingsIcon />
          </IconButton>
        </CareIntervalWrapper>
      </BarWrapper>
      <BarWrapper>
        <Bar>
          <FertilizerIndicator
            data-testid="fertilizer-indicator"
            weeksLeft={plant.weeksLeft}
            fertilizerInterval={plant.fertilizerInterval}
          />
        </Bar>
        <CareIntervalWrapper>
          <Text>
            Fertilizer {plant.weeksLeft}/{plant.fertilizerInterval} weeks left
          </Text>
          <IconButton
            data-testid="fertilizer-button"
            name="fertilizer"
            onClick={openCareSettings}
          >
            <SettingsIcon />
          </IconButton>
        </CareIntervalWrapper>
      </BarWrapper>
    </Container>
  )

  function openCareSettings(event) {
    setShowCareSetting(event.currentTarget.name)
  }
}

const Container = styled.div`
  display: grid;
  gap: 20px;
  padding: 10px 0 0;
  width: 100%;
`

const BarWrapper = styled.div`
  display: grid;
  gap: 10px;
`

const Bar = styled.div`
  background-color: var(--bar-bg-color);
  border-radius: 2px;
  height: 4px;
  position: relative;
  width: 100%;
`

const Indicator = styled.div`
  border-radius: 4px;
  height: 8px;
  min-width: 10px;
  position: absolute;
  top: -2px;
`

const WaterIndicator = styled(Indicator)`
  background-color: ${(props) =>
    props.daysLeft < 2 ? 'var(--warning-color)' : 'var(--bar-water-color)'};
  width: ${(props) => (100 / props.waterInterval) * props.daysLeft}%;
`

const FertilizerIndicator = styled(Indicator)`
  background-color: ${(props) =>
    props.weeksLeft < 2
      ? 'var(--warning-color)'
      : 'var(--bar-fertilizer-color)'};
  width: ${(props) => (100 / props.fertilizerInterval) * props.weeksLeft}%;
`

const CareIntervalWrapper = styled.div`
  display: flex;
  justify-content: space-between;
`

const Text = styled.p`
  color: var(--secondary-plant-font-color);
  font-size: 1rem;
`

const IconButton = styled.button`
  background-color: transparent;
  border: none;
`
