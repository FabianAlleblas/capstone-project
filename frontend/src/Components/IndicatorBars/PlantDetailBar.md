PlantDetailBar: normal

```js
<PlantDetailBar
  plant={{
    daysLeft: 10,
    weeksLeft: 4,
    waterInterval: 10,
    fertilizerInterval: 4,
  }}
  setShowCareSetting={() => {}}
/>
```

PlantDetailBar: low

```js
<PlantDetailBar
  plant={{
    daysLeft: 1,
    weeksLeft: 1,
    waterInterval: 10,
    fertilizerInterval: 4,
  }}
  setShowCareSetting={() => {}}
/>
```
