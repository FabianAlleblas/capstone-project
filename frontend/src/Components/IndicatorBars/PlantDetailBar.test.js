import { render } from '@testing-library/react'
import 'jest-styled-components'
import user from '@testing-library/user-event'
import PlantDetailBar from './PlantDetailBar'

const plant = {
  daysLeft: 10,
  weeksLeft: 4,
  waterInterval: 10,
  fertilizerInterval: 4,
}

const setShowCareSettingMock = jest.fn()

describe('PlantDetailBar', () => {
  it('renders correctly', () => {
    const { container } = render(
      <PlantDetailBar
        plant={plant}
        setShowCareSetting={setShowCareSettingMock}
      />
    )
    expect(container.firstChild).toMatchSnapshot()
  })
  it('calls setShowCareSettingMock with the correct name', () => {
    const { getByTestId } = render(
      <PlantDetailBar
        plant={plant}
        setShowCareSetting={setShowCareSettingMock}
      />
    )
    const WaterIconButton = getByTestId('water-button')

    user.click(WaterIconButton)

    expect(setShowCareSettingMock).not.toHaveBeenCalledWith('fertilizer')
    expect(setShowCareSettingMock).toHaveBeenCalledWith('water')
  })
  it('renders the full length of the bars', () => {
    const { getByTestId } = render(
      <PlantDetailBar
        plant={plant}
        setShowCareSetting={setShowCareSettingMock}
      />
    )
    const WaterBar = getByTestId('water-indicator')
    const FertilizerBar = getByTestId('fertilizer-indicator')

    expect(WaterBar).toHaveStyle('width: 100%;')
    expect(FertilizerBar).toHaveStyle('width: 100%')
  })
  it('renders the min length of the bars', () => {
    const { getByTestId } = render(
      <PlantDetailBar
        plant={{
          daysLeft: 0,
          weeksLeft: 0,
          waterInterval: 10,
          fertilizerInterval: 4,
        }}
        setShowCareSetting={setShowCareSettingMock}
      />
    )
    const WaterBar = getByTestId('water-indicator')
    const FertilizerBar = getByTestId('fertilizer-indicator')

    expect(WaterBar).toHaveStyle('width: 0%')
    expect(FertilizerBar).toHaveStyle('width: 0%')
  })
})
