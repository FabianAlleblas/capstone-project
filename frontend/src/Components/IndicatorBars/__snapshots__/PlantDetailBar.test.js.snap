// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantDetailBar renders correctly 1`] = `
.c0 {
  display: grid;
  gap: 20px;
  padding: 10px 0 0;
  width: 100%;
}

.c1 {
  display: grid;
  gap: 10px;
}

.c2 {
  background-color: var(--bar-bg-color);
  border-radius: 2px;
  height: 4px;
  position: relative;
  width: 100%;
}

.c3 {
  border-radius: 4px;
  height: 8px;
  min-width: 10px;
  position: absolute;
  top: -2px;
}

.c4 {
  background-color: var(--bar-water-color);
  width: 100%;
}

.c8 {
  background-color: var(--bar-fertilizer-color);
  width: 100%;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c6 {
  color: var(--secondary-plant-font-color);
  font-size: 1rem;
}

.c7 {
  background-color: transparent;
  border: none;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3 c4"
        data-testid="water-indicator"
      />
    </div>
    <div
      class="c5"
    >
      <p
        class="c6"
      >
        Water 
        10
        /
        10
         days left
      </p>
      <button
        class="c7"
        data-testid="water-button"
        name="water"
      >
        <svg>
          settingsicon.svg
        </svg>
      </button>
    </div>
  </div>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3 c8"
        data-testid="fertilizer-indicator"
      />
    </div>
    <div
      class="c5"
    >
      <p
        class="c6"
      >
        Fertilizer 
        4
        /
        4
         weeks left
      </p>
      <button
        class="c7"
        data-testid="fertilizer-button"
        name="fertilizer"
      >
        <svg>
          settingsicon.svg
        </svg>
      </button>
    </div>
  </div>
</div>
`;
